File: golang/sort/bucketSort.go
package sort

// 排序名称：桶排序
// 基本原理：分配N个桶, 遍历数组，数字放进对应的桶里, 然后再从桶中将数字按顺序拿出来形成一个有序数组
// 时间复杂度: O(M+N), M为桶数量, N为数组长度
// 空间复杂度: O(M)

// 实现一个简易的桶排序
func BucketSort(array []int) {
	var bucket [255]int
	for _, val := range array {
		bucket[val]++
	}

	for i,k := 0, 0; i < 255; i=i+1 {
		for j := 0; j < bucket[i]; j,k=j+1, k+1 {
			array[k] = i
		}
	}
}

File: javaSrc/src/sort/bucketSort.java
package sort;

import java.util.Random;

// 功能: 桶排序
// 基本思想: 遍历数组,讲数组存入有序的桶中, 记录每个桶出现的次数, 然后再遍历一遍桶
// 时间复杂度:
// 空间复杂度: O(N)
public class bucketSort {
    public static void sort(int[] array) {
        int[] bucket = new int[100000];

        // 遍历数组, 放入桶中
        for(int i=0; i < array.length; i++) {
            bucket[array[i]]++;
        }

        // 遍历桶, 拿出有序数据
        for(int i=0, j=0; i < 100000; i++) {
            while(bucket[i] > 0) {
                array[j] = i;
                j++;
                bucket[i]--;
            }
        }
    }

    public static void main(String[] args) {
        int[] array = new int[100000];

        Random random = new Random();
        for(int i=0; i<100000; i++) {
            array[i] = random.nextInt(100000);
        }
        sort(array);
        for(int i : array) {
            System.out.println(i);
        }
    }
}
